src/Components/Components/src/Virtualization/InfiniteScroll.cs

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.Components.Virtualization
+{
+    /// <summary>
+    /// Provides functionality for rendering a virtualized, infinite list of items.
+    /// </summary>
+    /// <typeparam name="TItem">The <c>context</c> type for the items being rendered.</typeparam>
+    public sealed class InfiniteScroll<TItem> : VirtualizeBase<TItem>
```

Comment:
I'm definitely open to distinguishing the `<Virtualize>` component from `<InfiniteScroll>`. You've pointed out some advantages in terms of having fewer possible invalid combinations of parameters.

It would be great to fully understand the tradeoffs we'd hit if we considered an alternative approach where infinite scrolling was purely a usage pattern within a single `<Virtualize>` component.

 * If you specify an `ItemHeight` (or `ItemSize`) then we automatically do random-access seeking to arbitrary points in the set as the user scrolls, and render a spacer-after that expands to whatever height is needed to make the scrollbar represent what we know about the full but finite scroll range. Presumably each time the developer fetches some more data, they could choose to update an `ItemCount` on the `<Virtualize>`, which would update the effective scroll range.
 * Or if you don't specify `ItemHeight`, then the spacer-after is only as high as it naturally is according to a `<SpacerAfter>` template parameter provided by the developer, so in effect there's no random access because there's no way to scroll further than the `<SpacerAfter>` element. It's up to the developer to fetch as many additional items as they want each time that becomes visible, and optionally signal that they've run out of items (maybe by updating `ItemCount` to match the total already fetched) at which point we'd no longer render the spacer-after.

This is not to say I'm fixated on this alternative approach. The approach you already have here might be better, but I want to be explicit about the tradeoffs :) The functionality we ship in `M.A.Components` needs to remain supported as-is forever so we feel uncertain about parts of it, I'd err on the side of simply skipping those pieces (e.g., infinite scrolling) until we get more confident in the future.

Reply:
I've done my best to implement what you've suggested here, albeit with some differences - most notably the omission of `ItemCount`. I tried the idea of using `ItemCount` to hint to `Virtualize` whether there are more items to fetch, but from a usability standpoint, I personally found it to be a bit unintuitive (often the `ItemCount` provided was just the number of items fetched, plus 1, and removing that plus 1 when there were no items remaining complicated the code). I also found that it was difficult to make scrolling behave correctly when `ItemCount` was larger than item count of the internal collection. Perhaps there is a nice solution to this - I didn't put in too much time investigating this specific problem. If the idea was to allow users to scroll freely without getting stuck waiting for items to load, I think this should be feasible by generating placeholder items using `ItemsProvider`, and replacing them after the asynchronous "fetching" operation completes. Please let me know what your thoughts are on this, since there might be some use cases I haven't yet considered.